File: src/performance_monitor.py
import time


class PerformanceMonitor:

    def __init__(self):
        self.start_times = {}
        self.elapsed_times = {}
        self.display_timer = time.time()

    def start(self, stage):
        self.start_times[stage] = time.perf_counter()

    def stop(self, stage):
        if stage in self.start_times:
            elapsed = (time.perf_counter() - self.start_times[stage]) * 1000
            self.elapsed_times[stage] = elapsed

    def get(self, stage):
        return self.elapsed_times.get(stage, 0.0)

    def total(self):
        return sum(self.elapsed_times.values())

    def reset(self):
        self.elapsed_times.clear()


def display_performance(self):

    current_time = time.time()

    if current_time - self.display_timer >= 1.0:

        print("--------------------------------")
        print(f"Capture          : {self.get('capture'):.2f} ms")
        print(f"FaceMesh         : {self.get('face_mesh'):.2f} ms")
        print(f"EyeMonitor       : {self.get('eye_monitor'):.2f} ms")
        print(f"AttentionMonitor : {self.get('attention_monitor'):.2f} ms")
        print(f"Total            : {self.total():.2f} ms")

        self.display_timer = current_time

File: cpp/src/performance_monitor.cpp
#include "performance_monitor.hpp"

#include <iostream>
#include <iomanip>

namespace dms
{

PerformanceMonitor::PerformanceMonitor()
{
    displayTimer = Clock::now();
}

void PerformanceMonitor::start(const std::string& stage)
{
    startTimes[stage] = Clock::now();
}

void PerformanceMonitor::stop(const std::string& stage)
{
    auto it = startTimes.find(stage);

    if (it != startTimes.end())
    {
        auto elapsed =
            std::chrono::duration<double, std::milli>(
                Clock::now() - it->second).count();

        elapsedTimes[stage] = elapsed;
    }
}

double PerformanceMonitor::get(const std::string& stage) const
{
    auto it = elapsedTimes.find(stage);

    if (it != elapsedTimes.end())
        return it->second;

    return 0.0;
}

double PerformanceMonitor::total() const
{
    double sum = 0.0;

    for (const auto& item : elapsedTimes)
        sum += item.second;

    return sum;
}

void PerformanceMonitor::reset()
{
    elapsedTimes.clear();
}

void PerformanceMonitor::displayPerformance()
{
    auto currentTime = Clock::now();

    auto elapsed =
        std::chrono::duration<double>(
            currentTime - displayTimer).count();

    if (elapsed >= 1.0)
    {
        std::cout << "--------------------------------\n";

        std::cout << std::fixed << std::setprecision(2);

        std::cout << "Capture          : "
                  << get("capture")
                  << " ms\n";

        std::cout << "FaceMesh         : "
                  << get("face_mesh")
                  << " ms\n";

        std::cout << "EyeMonitor       : "
                  << get("eye_monitor")
                  << " ms\n";

        std::cout << "AttentionMonitor : "
                  << get("attention_monitor")
                  << " ms\n";

        std::cout << "Total            : "
                  << total()
                  << " ms\n";

        displayTimer = currentTime;
    }
}

}   // namespace dms